### PR TITLE
[Symbols] Add support for the new Xcode 15 Symbols framework.

### PR DESCRIPTION
**Pull request by mandel-macaque**

### CODE REVIEW

**Comment by mandel-macaque** [comment]
Leaving as draft until we land the xcode15 support. But good to keep the record straight.

File: src/symbols.cs [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+	[BaseType (typeof (NSObject))]
+	interface NSSymbolContentTransition : NSCopying, NSSecureCoding { }
+
+	[Watch (10, 0), TV (17, 0), Mac (14, 0), iOS (17, 0), MacCatalyst (17, 0)]
+	[BaseType (typeof (NSObject))]
+	[DisableDefaultCtor]
+	interface NSSymbolEffectOptions : NSCopying, NSSecureCoding {
+		[Static]
+		[Export ("options")]
+		NSSymbolEffectOptions CreateOptions ();
```
Comment (rolfbjarne) [inline src/symbols.cs:22]: I think it sounds better if you remove `Options`:

```suggestion
		NSSymbolEffectOptions Create ();
```

File: src/symbols.cs [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+
+		[Static]
+		[Export ("optionsWithRepeating")]
+		NSSymbolEffectOptions CreateOptionsWithRepeating ();
```
Comment (rolfbjarne) [inline src/symbols.cs:26]: Same, but also removing `With`:

```suggestion
		NSSymbolEffectOptions CreateRepeating ();
```

and the same for the other Get and Create methods here.

This matches the Swift names better (and also sounds better).

File: src/symbols.cs [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+
+	[Watch (10, 0), TV (17, 0), Mac (14, 0), iOS (17, 0), MacCatalyst (17, 0)]
+	[BaseType (typeof (NSSymbolEffect))]
+	interface NSSymbolPulseEffect {
+		[Static]
+		[Export ("effect")]
+		NSSymbolPulseEffect CreateEffect ();
+
+		[Export ("effectWithByLayer")]
+		NSSymbolPulseEffect GetEffectWithByLayer ();
```
Comment (rolfbjarne) [inline src/symbols.cs:61]: These are properties in Swift (https://developer.apple.com/documentation/symbols/pulsesymboleffect/4197828-bylayer)

```suggestion
		NSSymbolPulseEffect ByLayer { get; }
```

File: src/symbols.cs [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+
+		[Static]
+		[Export ("bounceDownEffect")]
+		NSSymbolBounceEffect CreateBounceDownEffect ();
+
+		[Export ("effectWithByLayer")]
+		NSSymbolBounceEffect GetEffectWithByLayer ();
+
+		[Export ("effectWithWholeSymbol")]
+		NSSymbolBounceEffect GetEffectWithWholeSymbol ();
```
Comment (rolfbjarne) [inline src/symbols.cs:86]: Same, these can be properties and named better.

File: src/symbols.cs [hunk 5]
```diff
@@ -0,0 +1,215 @@
...
+
+	[Watch (10, 0), TV (17, 0), Mac (14, 0), iOS (17, 0), MacCatalyst (17, 0)]
+	[BaseType (typeof (NSSymbolEffect))]
+	interface NSSymbolVariableColorEffect {
+		[Static]
+		[Export ("effect")]
+		NSSymbolVariableColorEffect CreateEffect ();
+
+		[Export ("effectWithIterative")]
+		NSSymbolVariableColorEffect GetEffectWithIterative ();
```
Comment (rolfbjarne) [inline src/symbols.cs:97]: Same, simplified and property:

```suggestion
		NSSymbolVariableColorEffect Iterative { get; }
```

File: src/symbols.cs [hunk 6]
```diff
@@ -0,0 +1,215 @@
...
+		[Export ("effectWithWholeSymbol")]
+		NSSymbolBounceEffect GetEffectWithWholeSymbol ();
+	}
+
+	[Watch (10, 0), TV (17, 0), Mac (14, 0), iOS (17, 0), MacCatalyst (17, 0)]
+	[BaseType (typeof (NSSymbolEffect))]
+	interface NSSymbolVariableColorEffect {
+		[Static]
+		[Export ("effect")]
+		NSSymbolVariableColorEffect CreateEffect ();
```
Comment (rolfbjarne) [inline src/symbols.cs:94]:
```suggestion
		NSSymbolVariableColorEffect Create ();
```

File: src/symbols.cs [hunk 7]
```diff
@@ -0,0 +1,215 @@
...
+		[Export ("effectWithDimInactiveLayers")]
+		NSSymbolVariableColorEffect GetEffectWithDimInactiveLayers ();
+	}
+
+	[Watch (10, 0), TV (17, 0), Mac (14, 0), iOS (17, 0), MacCatalyst (17, 0)]
+	[BaseType (typeof (NSSymbolEffect))]
+	interface NSSymbolScaleEffect {
+		[Static]
+		[Export ("effect")]
+		NSSymbolScaleEffect CreateEffect ();
```
Comment (rolfbjarne) [inline src/symbols.cs:120]:
```suggestion
		NSSymbolScaleEffect Create ();
```

File: src/symbols.cs [hunk 8]
```diff
@@ -0,0 +1,215 @@
...
+
+		[Static]
+		[Export ("scaleDownEffect")]
+		NSSymbolScaleEffect CreateScaleDownEffect ();
+
+		[Export ("effectWithByLayer")]
+		NSSymbolScaleEffect GetEffectWithByLayer ();
+
+		[Export ("effectWithWholeSymbol")]
+		NSSymbolScaleEffect GetEffectWithWholeSymbol ();
```
Comment (rolfbjarne) [inline src/symbols.cs:134]: Simplified properties

File: src/symbols.cs [hunk 9]
```diff
@@ -0,0 +1,215 @@
...
+		[Export ("transitionWithWholeSymbol")]
+		NSSymbolReplaceContentTransition GetTransitionWithWholeSymbol ();
+	}
+
+	[Watch (10, 0), TV (17, 0), Mac (14, 0), iOS (17, 0), MacCatalyst (17, 0)]
+	[BaseType (typeof (NSSymbolContentTransition))]
+	interface NSSymbolAutomaticContentTransition {
+		[Static]
+		[Export ("transition")]
+		NSSymbolAutomaticContentTransition CreateTransition ();
```
Comment (rolfbjarne) [inline src/symbols.cs:212]: I stopped suggesting, they all follow the same pattern :)

**Review by rolfbjarne — CHANGES_REQUESTED** [review]

File: src/symbols.cs [hunk 10]
```diff
@@ -0,0 +1,215 @@
...
+	[BaseType (typeof (NSObject))]
+	[DisableDefaultCtor]
+	interface NSSymbolEffectOptions : NSCopying, NSSecureCoding {
+		[Static]
+		[Export ("options")]
+		NSSymbolEffectOptions Create ();
+
+		[Static]
+		[Export ("optionsWithRepeating")]
+		NSSymbolEffectOptions CreatehRepeating ();
```
Comment (rolfbjarne) [inline src/symbols.cs:26]:
```suggestion
		NSSymbolEffectOptions CreateRepeating ();
```

File: src/symbols.cs [hunk 11]
```diff
@@ -0,0 +1,215 @@
...
+
+		[Export ("optionsWithRepeating")]
+		NSSymbolEffectOptions GetWithRepeating ();
```
Comment (rolfbjarne) [inline src/symbols.cs:29]: All the methods follow the same pattern and need the same change...

```suggestion
		NSSymbolEffectOptions GetRepeating ();
```

File: src/symbols.cs [hunk 12]
```diff
@@ -0,0 +1,215 @@
...
+
+	[Watch (10, 0), TV (17, 0), Mac (14, 0), iOS (17, 0), MacCatalyst (17, 0)]
+	[BaseType (typeof (NSSymbolEffect))]
+	interface NSSymbolPulseEffect {
+		[Static]
+		[Export ("effect")]
+		NSSymbolPulseEffect Create ();
+
+		[Export ("effectWithByLayer")]
+		NSSymbolPulseEffect ByLager { get; }
```
Comment (rolfbjarne) [inline src/symbols.cs:61]:
```suggestion
		NSSymbolPulseEffect ByLayer { get; }
```

File: src/symbols.cs [hunk 13]
```diff
@@ -0,0 +1,215 @@
...
+		[Static]
+		[Export ("bounceUpEffect")]
+		NSSymbolBounceEffect CreateBounceUpEffect ();
+
+		[Static]
+		[Export ("bounceDownEffect")]
+		NSSymbolBounceEffect CreateBounceDownEffect ();
+
+		[Export ("effectWithByLayer")]
+		NSSymbolBounceEffect ByLager { get; }
```
Comment (rolfbjarne) [inline src/symbols.cs:83]:
```suggestion
		NSSymbolBounceEffect ByLayer { get; }
```

File: src/symbols.cs [hunk 14]
```diff
@@ -0,0 +1,215 @@
...
+		[Static]
+		[Export ("scaleUpEffect")]
+		NSSymbolScaleEffect CreateScaleUpEffect ();
+
+		[Static]
+		[Export ("scaleDownEffect")]
+		NSSymbolScaleEffect CreateScaleDownEffect ();
+
+		[Export ("effectWithByLayer")]
+		NSSymbolScaleEffect ByLager { get; }
```
Comment (rolfbjarne) [inline src/symbols.cs:131]:
```suggestion
		NSSymbolScaleEffect ByLayer { get; }
```

File: src/symbols.cs [hunk 15]
```diff
@@ -0,0 +1,215 @@
...
+		[Static]
+		[Export ("appearUpEffect")]
+		NSSymbolAppearEffect CreateAppearUpEffect ();
+
+		[Static]
+		[Export ("appearDownEffect")]
+		NSSymbolAppearEffect CreateAppearDownEffect ();
+
+		[Export ("effectWithByLayer")]
+		NSSymbolAppearEffect ByLager { get; }
```
Comment (rolfbjarne) [inline src/symbols.cs:153]:
```suggestion
		NSSymbolAppearEffect ByLayer { get; }
```

File: src/symbols.cs [hunk 16]
```diff
@@ -0,0 +1,215 @@
...
+		[Static]
+		[Export ("disappearUpEffect")]
+		NSSymbolDisappearEffect CreateDisappearUpEffect ();
+
+		[Static]
+		[Export ("disappearDownEffect")]
+		NSSymbolDisappearEffect CreateDisappearDownEffect ();
+
+		[Export ("effectWithByLayer")]
+		NSSymbolDisappearEffect ByLager { get; }
```
Comment (rolfbjarne) [inline src/symbols.cs:175]:
```suggestion
		NSSymbolDisappearEffect ByLayer { get; }
```

File: src/symbols.cs [hunk 17]
```diff
@@ -0,0 +1,215 @@
...
+		[Static]
+		[Export ("replaceUpUpTransition")]
+		NSSymbolReplaceContentTransition CreateReplaceUpUpTransition ();
+
+		[Static]
+		[Export ("replaceOffUpTransition")]
+		NSSymbolReplaceContentTransition CreateReplaceOffUpTransition ();
+
+		[Export ("transitionWithByLayer")]
+		NSSymbolReplaceContentTransition ByLager { get; }
```
Comment (rolfbjarne) [inline src/symbols.cs:201]:
```suggestion
		NSSymbolReplaceContentTransition ByLayer { get; }
```

🍻

Comment (mandel-macaque) [reply src/symbols.cs:201]: Guess who got a typo in the search and replace 😅

File: src/symbols.cs [hunk 18]
```diff
@@ -0,0 +1,215 @@
...
+		[Export ("effectWithWholeSymbol")]
+		NSSymbolDisappearEffect WholeSymbol { get; }
+	}
+
+	[Watch (10, 0), TV (17, 0), Mac (14, 0), iOS (17, 0), MacCatalyst (17, 0)]
+	[BaseType (typeof (NSSymbolContentTransition))]
+	interface NSSymbolReplaceContentTransition {
+		[Static]
+		[Export ("transition")]
+		NSSymbolReplaceContentTransition CreateTransition ();
```
Comment (rolfbjarne) [inline src/symbols.cs:186]:
```suggestion
		NSSymbolReplaceContentTransition Create ();
```

**Review by rolfbjarne — CHANGES_REQUESTED** [review]

File: src/symbols.cs [hunk 19]
```diff
@@ -0,0 +1,215 @@
...
+		[Static]
+		[Export ("optionsWithRepeating")]
+		NSSymbolEffectOptions CreateRepeating ();
+
+		[Export ("optionsWithRepeating")]
+		NSSymbolEffectOptions GetRepeating ();
+
+		[Static]
+		[Export ("optionsWithNonRepeating")]
+		NSSymbolEffectOptions CreatesWithNonRepeating ();
```
[the remark this answers is not in the document]
Comment (rolfbjarne) [reply src/symbols.cs:33]: Did you forget this one?

**Review by rolfbjarne — APPROVED** [review]

File: src/symbols.cs [hunk 20]
```diff
@@ -0,0 +1,215 @@
+using CoreFoundation;
+using ObjCRuntime;
+using Foundation;
+
+using System;
+
+namespace Symbols {
+	[Watch (10, 0), TV (17, 0), Mac (14, 0), iOS (17, 0), MacCatalyst (17, 0)]
+	[BaseType (typeof (NSObject))]
+	interface NSSymbolEffect : NSCopying, NSSecureCoding { }
```
Comment (dalexsoto) [inline src/symbols.cs:10]: Documentation says it should not be used.

```suggestion
	[BaseType (typeof (NSObject))]
	[DisableDefaultCtor]
	interface NSSymbolEffect : NSCopying, NSSecureCoding { }
```

File: src/symbols.cs [hunk 21]
```diff
@@ -0,0 +1,215 @@
...
+		[Export ("optionsWithSpeed:")]
+		NSSymbolEffectOptions Create (double speed);
+
+		[Export ("optionsWithSpeed:")]
+		NSSymbolEffectOptions Get (double speed);
+	}
+
+	[Watch (10, 0), TV (17, 0), Mac (14, 0), iOS (17, 0), MacCatalyst (17, 0)]
+	[BaseType (typeof (NSSymbolEffect))]
+	interface NSSymbolPulseEffect {
```
Comment (dalexsoto) [inline src/symbols.cs:55]: There is a create.

```suggestion
	[BaseType (typeof (NSSymbolEffect))]
	[DisableDefaultCtor]
	interface NSSymbolPulseEffect {
```

File: src/symbols.cs [hunk 22]
```diff
@@ -0,0 +1,215 @@
...
+		[Export ("effectWithByLayer")]
+		NSSymbolPulseEffect ByLayer { get; }
+
+		[Export ("effectWithWholeSymbol")]
+		NSSymbolPulseEffect WholeSymbol { get; }
+	}
+
+	[Watch (10, 0), TV (17, 0), Mac (14, 0), iOS (17, 0), MacCatalyst (17, 0)]
+	[BaseType (typeof (NSSymbolEffect))]
+	interface NSSymbolBounceEffect {
```
Comment (dalexsoto) [inline src/symbols.cs:69]: Same

```suggestion
	[BaseType (typeof (NSSymbolEffect))]
	[DisableDefaultCtor]
	interface NSSymbolBounceEffect {
```

File: src/symbols.cs [hunk 23]
```diff
@@ -0,0 +1,215 @@
...
+		[Export ("effectWithByLayer")]
+		NSSymbolBounceEffect ByLayer { get; }
+
+		[Export ("effectWithWholeSymbol")]
+		NSSymbolBounceEffect WholeSymbol { get; }
+	}
+
+	[Watch (10, 0), TV (17, 0), Mac (14, 0), iOS (17, 0), MacCatalyst (17, 0)]
+	[BaseType (typeof (NSSymbolEffect))]
+	interface NSSymbolVariableColorEffect {
```
Comment (dalexsoto) [inline src/symbols.cs:91]: Same
```suggestion
	[BaseType (typeof (NSSymbolEffect))]
	[DisableDefaultCtor]
	interface NSSymbolVariableColorEffect {
```

File: src/symbols.cs [hunk 24]
```diff
@@ -0,0 +1,215 @@
...
+		[Export ("effectWithHideInactiveLayers")]
+		NSSymbolVariableColorEffect HideInactiveLayers { get; }
+
+		[Export ("effectWithDimInactiveLayers")]
+		NSSymbolVariableColorEffect DimInactiveLayers { get; }
+	}
+
+	[Watch (10, 0), TV (17, 0), Mac (14, 0), iOS (17, 0), MacCatalyst (17, 0)]
+	[BaseType (typeof (NSSymbolEffect))]
+	interface NSSymbolScaleEffect {
```
Comment (dalexsoto) [inline src/symbols.cs:117]: Same
```suggestion
	[BaseType (typeof (NSSymbolEffect))]
	[DisableDefaultCtor]
	interface NSSymbolScaleEffect {
```

File: src/symbols.cs [hunk 25]
```diff
@@ -0,0 +1,215 @@
...
+		[Export ("effectWithByLayer")]
+		NSSymbolScaleEffect ByLayer { get; }
+
+		[Export ("effectWithWholeSymbol")]
+		NSSymbolScaleEffect WholeSymbol { get; }
+	}
+
+	[Watch (10, 0), TV (17, 0), Mac (14, 0), iOS (17, 0), MacCatalyst (17, 0)]
+	[BaseType (typeof (NSSymbolEffect))]
+	interface NSSymbolAppearEffect {
```
Comment (dalexsoto) [inline src/symbols.cs:139]:
```suggestion
	[BaseType (typeof (NSSymbolEffect))]
	[DisableDefaultCtor]
	interface NSSymbolAppearEffect {
```

File: src/symbols.cs [hunk 26]
```diff
@@ -0,0 +1,215 @@
...
+		[Export ("effectWithByLayer")]
+		NSSymbolAppearEffect ByLayer { get; }
+
+		[Export ("effectWithWholeSymbol")]
+		NSSymbolAppearEffect WholeSymbol { get; }
+	}
+
+	[Watch (10, 0), TV (17, 0), Mac (14, 0), iOS (17, 0), MacCatalyst (17, 0)]
+	[BaseType (typeof (NSSymbolEffect))]
+	interface NSSymbolDisappearEffect {
```
Comment (dalexsoto) [inline src/symbols.cs:161]:
```suggestion
	[BaseType (typeof (NSSymbolEffect))]
	[DisableDefaultCtor]
	interface NSSymbolDisappearEffect {
```

File: src/symbols.cs [hunk 27]
```diff
@@ -0,0 +1,215 @@
...
+using System;
+
+namespace Symbols {
+	[Watch (10, 0), TV (17, 0), Mac (14, 0), iOS (17, 0), MacCatalyst (17, 0)]
+	[BaseType (typeof (NSObject))]
+	interface NSSymbolEffect : NSCopying, NSSecureCoding { }
+
+	[Watch (10, 0), TV (17, 0), Mac (14, 0), iOS (17, 0), MacCatalyst (17, 0)]
+	[BaseType (typeof (NSObject))]
+	interface NSSymbolContentTransition : NSCopying, NSSecureCoding { }
```
Comment (dalexsoto) [inline src/symbols.cs:14]: Docs say that subclasses are the ones meant to be used.
```suggestion
	[BaseType (typeof (NSObject))]
	[DisableDefaultCtor]
	interface NSSymbolContentTransition : NSCopying, NSSecureCoding { }
```

File: src/symbols.cs [hunk 28]
```diff
@@ -0,0 +1,215 @@
...
+		[Export ("effectWithByLayer")]
+		NSSymbolDisappearEffect ByLayer { get; }
+
+		[Export ("effectWithWholeSymbol")]
+		NSSymbolDisappearEffect WholeSymbol { get; }
+	}
+
+	[Watch (10, 0), TV (17, 0), Mac (14, 0), iOS (17, 0), MacCatalyst (17, 0)]
+	[BaseType (typeof (NSSymbolContentTransition))]
+	interface NSSymbolReplaceContentTransition {
```
Comment (dalexsoto) [inline src/symbols.cs:183]: same
```suggestion
	[BaseType (typeof (NSSymbolContentTransition))]
	[DisableDefaultCtor]
	interface NSSymbolReplaceContentTransition {
```

File: src/symbols.cs [hunk 29]
```diff
@@ -0,0 +1,215 @@
...
+		[Export ("transitionWithByLayer")]
+		NSSymbolReplaceContentTransition ByLayer { get; }
+
+		[Export ("transitionWithWholeSymbol")]
+		NSSymbolReplaceContentTransition WholeSymbol { get; }
+	}
+
+	[Watch (10, 0), TV (17, 0), Mac (14, 0), iOS (17, 0), MacCatalyst (17, 0)]
+	[BaseType (typeof (NSSymbolContentTransition))]
+	interface NSSymbolAutomaticContentTransition {
```
Comment (dalexsoto) [inline src/symbols.cs:209]: Same
```suggestion
	[BaseType (typeof (NSSymbolContentTransition))]
	[DisableDefaultCtor]
	interface NSSymbolAutomaticContentTransition {
```

**Review by dalexsoto — CHANGES_REQUESTED** [review]
Nice 👍

**Review by dalexsoto — APPROVED** [review]

**Pull request merged**